### PR TITLE
feat: further split the smallest possible set to compact but too large to do so

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Forth PR and the most important step for https://github.com/influxdata/idpe/issues/17246

### CODE REVIEW

File: compactor2/tests/layouts/large_files.rs [hunk 1]
```diff
@@ -283,19 +278,57 @@ async fn two_large_files_total_over_max_compact_size() {
...
+    - "Committing partition 1:"
+    - "  Soft Deleting 2 files: L1.1, L2.2"
+    - "  Creating 4 files"
+    - "**** Simulation run 2, type=split(split_times=[668]). 1 Input Files, 50mb total:"
+    - "L1, all files 50mb                                                                                                 "
+    - "L1.4[668,1000] 1ns       |------------------------------------------L1.4------------------------------------------|"
+    - "**** 2 Output Files (parquet_file_id not yet assigned), 50mb total:"
+    - "L1                                                                                                                 "
+    - "L1.?[668,668] 1ns 0b     |L1.?|                                                                                    "
+    - "L1.?[669,1000] 1ns 50mb  |-----------------------------------------L1.?------------------------------------------| "
```
Comment (NGA-TRAN) [inline compactor2/tests/layouts/large_files.rs:304]: This is the case that we should allow to split L1 with range [668, 1000] into 2 files [668, 668] and [669, 1000] because we have two L2 files with ranges [2,668] and [669, 1000]. By spitting L1 like that, we will guarantee the new L1 files each only overlap with at most one L2 file.

We will throw error and skip  the partition if we need to split tiny time range further. In this case we do not and it should work.

File: compactor2_test_utils/src/commit_wrapper.rs [hunk 2]
```diff
@@ -63,11 +63,7 @@ impl Commit for CommitRecorder {
             }
 
             if !create.is_empty() {
-                run_log.push(format!(
-                    "  Creating {} files at level {}",
-                    create.len(),
-                    target_level
-                ));
+                run_log.push(format!("  Creating {} files", create.len()));
```
Comment (NGA-TRAN) [inline compactor2_test_utils/src/commit_wrapper.rs:66]: Since now a run can split 2 files in different levels, listing only one level here is misleading. So let us not list the target level of the run

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 3]
```diff
@@ -32,7 +32,7 @@ use crate::{
 /// To achieve this goal, a start-level file should be split to overlap with at most one target-level file. This enables the
 /// minimum set of compacting files to 2 files: a start-level file and an overlapped target-level file.
 ///
-pub fn identify_files_to_split(
+pub fn identify_start_level_files_to_split(
```
Comment (NGA-TRAN) [inline compactor2/src/components/split_or_compact/start_level_files_to_split.rs:35]: This file and its function are renamed to reflect its exact job: splitting start-level files. The new file `large_files_to_split.rs` is used to split the minimum set of compacting files in later step that after splitting start-level files and still not small enough to compact

File: compactor2/tests/layouts/core.rs [hunk 4]
```diff
@@ -103,7 +103,7 @@ async fn all_overlapping_l0() {
     - "L1.?[160021,200000] 10ns 18mb                                                                        |-----L1.?------| "
     - "Committing partition 1:"
     - "  Soft Deleting 10 files: L0.1, L0.2, L0.3, L0.4, L0.5, L0.6, L0.7, L0.8, L0.9, L0.10"
-    - "  Creating 2 files at level CompactionLevel::L1"
+    - "  Creating 2 files"
```
Comment (alamb) [inline compactor2/tests/layouts/core.rs:106]: Why did you remove this from the display? Is it because it is redundant with the `L1` output in `    - "**** 2 Output Files (parquet_file_id not yet assigned), 90mb total:"` above?

File: compactor2/tests/layouts/knobs.rs [hunk 5]
```diff
@@ -3450,7 +3450,7 @@ async fn all_overlapping_l0_max_input_bytes_per_partition_small_max_desired_file
     - "L1.?[185680,189405] 10ns 1.61mb                                                                        |-----L1.?------| "
     - "Committing partition 1:"
     - "  Soft Deleting 2 files: L0.302, L1.506"
-    - "  Creating 2 files at level CompactionLevel::L1"
+    - "  Creating 2 files"
```
Comment (alamb) [inline compactor2/tests/layouts/knobs.rs:3453]: FWIW you could make this diff significantly smaller (and thus easier to find the important differences) by creating a PR with just the output display changes first

File: compactor2/tests/layouts/large_files.rs [hunk 6]
```diff
@@ -9,7 +9,6 @@ use crate::layouts::{layout_setup_builder, parquet_builder, run_layout_scenario,
 const MAX_COMPACT_SIZE: usize = 300 * ONE_MB as usize;
 const MAX_DESIRED_FILE_SIZE: u64 = 100 * ONE_MB;
 
-// This file should be upgraded after https://github.com/influxdata/idpe/issues/17246
```
Comment (alamb) [inline compactor2/tests/layouts/large_files.rs:12]: Why should it not be upgraded now?

File: compactor2/tests/layouts/large_files.rs [hunk 7]
```diff
@@ -283,19 +278,57 @@ async fn two_large_files_total_over_max_compact_size() {
...
+    - "L2.5[2,668] 1ns 100.1mb  |-----------------------------------------L2.5------------------------------------------| "
+    - "**** 2 Output Files (parquet_file_id not yet assigned), 200.1mb total:"
+    - "L2                                                                                                                 "
+    - "L2.?[1,335] 1ns 100.2mb  |-------------------L2.?--------------------|                                             "
+    - "L2.?[336,668] 1ns 99.9mb                                              |-------------------L2.?-------------------| "
+    - "Committing partition 1:"
+    - "  Soft Deleting 3 files: L1.3, L2.5, L1.7"
+    - "  Creating 2 files"
+    - "**** Final Output Files "
+    - "L1                                                                                                                 "
```
Comment (alamb) [inline compactor2/tests/layouts/large_files.rs:322]: The final output looks good (better than skipped!) -- though it is sad there are overlapping L1 and L2 files.

There also seem to be twice as many rewrites (split plans) of the data than necessary Why can't the initial large files each be split with a single split plan with [335, 668] (resulting in three files) and then those output files be compacted?

If it is a limitation of how the current algorithm is structured, that makes sense though would be good to fix maybe as a follow on PR

File: compactor2/tests/layouts/large_files.rs [hunk 8]
```diff
@@ -283,19 +278,57 @@ async fn two_large_files_total_over_max_compact_size() {
...
+    - "L2.?[669,1000] 1ns 49.9mb                                                            |-----------L2.?------------| "
+    - "Committing partition 1:"
+    - "  Soft Deleting 2 files: L1.1, L2.2"
+    - "  Creating 4 files"
+    - "**** Simulation run 2, type=split(split_times=[668]). 1 Input Files, 50mb total:"
+    - "L1, all files 50mb                                                                                                 "
+    - "L1.4[668,1000] 1ns       |------------------------------------------L1.4------------------------------------------|"
+    - "**** 2 Output Files (parquet_file_id not yet assigned), 50mb total:"
+    - "L1                                                                                                                 "
+    - "L1.?[668,668] 1ns 0b     |L1.?|                                                                                    "
```
Comment (alamb) [inline compactor2/tests/layouts/large_files.rs:303]: This doesn't seem like a good plan -- to make a file with just a single time range [668 -- 668]

**Review by alamb — COMMENTED** [review]
Nice work @NGA-TRAN 

I didn't review and of the code changes, but I looked at the final output and the plans the compactor took to get there.

The final output looks mostly good 👍  (except that there are still some overlaps, but I believe that is a known limitation and will need something like "cold compaction") to handle

I had some questions on the plans the compactor takes, but basically I think this PR's behavior is an improvement. Since I didn't look at the algorithmic changes, I didn't approve the PR.

This PR will almost certainly conflict with https://github.com/influxdata/influxdb_iox/pull/7197

File: compactor2/tests/layouts/large_files.rs (hunk 6 above)
Comment (NGA-TRAN) [reply compactor2/tests/layouts/large_files.rs:12]: I just removed the comment. The output is upgraded and was done in a previous PR and already in main hence you do not see the changes here

File: compactor2/tests/layouts/large_files.rs (hunk 8 above)
Comment (NGA-TRAN) [reply compactor2/tests/layouts/large_files.rs:303]: See the [comment here](https://github.com/influxdata/influxdb_iox/pull/7180#discussion_r1134615267). It is actually a better plan

File: compactor2/tests/layouts/large_files.rs (hunk 7 above)
Comment (NGA-TRAN) [reply compactor2/tests/layouts/large_files.rs:322]: > though it is sad there are overlapping L1 and L2 files.

This is a hot compaction and leaving this overlap for coming data L0 to further compact is better than aggressively compacting it until no longer overlaps since there is still more data coming.

> There also seem to be twice as many rewrites (split plans) of the data than necessary Why can't the initial large files each be split with a single split plan with [335, 668] (resulting in three files) and then those output files be compacted?

The algorithm is done for common cases. When a very rare case like this happens, it won't be optimal. We are aiming for it to work first.

Here is the recursive algorithm for you to understand why:

If we hit over size limit, we do this in order and recursively:
1. Split all start-level files so each only overlaps with max one target-level file. This split will align the time ranges of many files. Joe and I think this is the most important and effective to do first in the order. 
2. Choose the smallest possible set to compact. Mostly after the split above, we will be able to do this because their time ranges are aligned
3. If the smallest possible set to compact is still over size limit, we will split over-size files. This will be any large files of  start-level or target-level. This split is very rare and the goal is to reduce the size for us to move forward, hence the split time will make each output file soft max file size.

Then when we recursively go over the loop. If still over size limit, step number 1 will split start-level files again to align time ranges with recent split in step number 3.

The short answer is if we have to do step1 after step 3, you will see that. And yes, you can say ` a limitation of how the current algorithm` but I would say in different way that the algorithm will work best for common cases and work ok for rare case like this

File: compactor2/tests/layouts/knobs.rs (hunk 5 above)
Comment (NGA-TRAN) [reply compactor2/tests/layouts/knobs.rs:3453]: This is a good idea 👍 

@Joe-Blount : since you will review this, let me know if you want me to make this change in a separate and commit it first or you are ok with this.

File: compactor2/tests/layouts/large_files.rs (hunk 7 above)
Comment (Joe-Blount) [reply compactor2/tests/layouts/large_files.rs:322]: >When a very rare case like this happens, it won't be optimal. We are aiming for it to work first.

The dashboards now show the rate of compact & split decisions.  Splits happen regularly in some clusters, but at a very low rate relative to compactions.  If this PR moves the graphs significantly we may need to re-evaluate, but the graphs will tell us.

image

Comment (NGA-TRAN) [reply compactor2/tests/layouts/large_files.rs:322]: This makes me think we may need to distinguish between split start-level files (step 1) with split large files (step 3) to decide what happens more and optimize as needed

File: compactor2/src/components/split_or_compact/files_to_compact.rs [hunk 9]
```diff
@@ -419,20 +516,59 @@ mod tests {
...
 
-        // There are only 3 choices for compacting:
-        //  1. Smallest set: L0.1 with L1.11
-        //  2. Medium size set: L0.1, L0.2 with L1.11, L1.12, L1.13
-        //  3. All files: L0.1, L0.2, L0.3 with L1.11, L1.12, L1.13
+        // There are only 4 choices:
+        //  1. Smallest set is still too large to compact. Split the set: L0.1 with L1.11
+        //  2. Smallest set to compact: L0.1 with L1.11
+        //  3. Medium size set to compact: L0.1, L0.2 with L1.11, L1.12, L1.13
+        //  4. All files to compact: L0.1, L0.2, L0.3 with L1.11, L1.12, L1.13
```
Comment (Joe-Blount) [inline compactor2/src/components/split_or_compact/files_to_compact.rs:523]: Just thinking out loud here, not necessarily suggesting change at this time...

I'm wondering if the split logic introduces the possibility of making it harder to catch up once we fall behind.  For example, if we typically don't do many splits, maybe because we're caught up and typically compact everything (choice 4), then if something causes us to fall behind so we pivot to choice 1.  Since it adds work, we might get stuck in mode where choice 1 makes us less efficient so we can't catch up.
I've been bitten by logic that adds work to do once you've fallen behind.  It can make it very difficult to catch back up and makes the system latch into a bad state it can't escape.  I don't have enough understanding of the typical ingest and compaction pattern to know if that's a possibility here.

I guess for now, we should just think about if as the compactor falls behind, does the split logic likely go to from a state of rarely being invoked to frequently being invoked?  If yes, that is a source of fragility in the system.

File: compactor2/src/components/split_or_compact/files_to_compact.rs [hunk 10]
```diff
@@ -456,14 +592,20 @@ mod tests {
         );
 
         // --------------------
-        // size limit = MAX_SIZE * 4 to force the first choice, L0.1 with L1.11, becasue it still not enough to for second choice
-        let (files_to_compact, files_to_keep) = limit_files_to_compact(
+        // size limit = MAX_SIZE * 4 to force the second choice, L0.1 with L1.11, becasue it still not enough to for second choice
```
Comment (Joe-Blount) [inline compactor2/src/components/split_or_compact/files_to_compact.rs:595]:
```suggestion
        // size limit = MAX_SIZE * 4 to force the second choice, L0.1 with L1.11, because it still not enough to for second choice
```

File: compactor2/src/components/split_or_compact/large_files_to_split.rs [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+    // Split files over max_desired_file_size into multiple files each is softly around max_desired_file_size
+    let mut files_to_split = Vec::with_capacity(files.len());
+    let mut files_not_to_split = Vec::with_capacity(files.len());
+    for file in files.into_iter() {
+        let file_size = file.file_size_bytes as u64;
+        let min_time = file.min_time.get();
+        let max_time = file.max_time.get();
+
+        // only split files that are larger than max_desired_file_size and have time range at least 2
+        if file_size > max_desired_file_size && file.min_time < file.max_time - 1 {
```
Comment (Joe-Blount) [inline compactor2/src/components/split_or_compact/large_files_to_split.rs:44]: Since `max_desired_file_size` limit is softly enforced, it is normal and ok for files to slightly exceed it.  
I'm concerned about this creating a routine pattern where we write files slightly exceeding `max_desired_file_size` then this code split them.  I wonder if we should have some arbitrary amount we allow the file to exceed `max_desired_file_size` to prevent excessive splitting?

**Comment by Joe-Blount** [comment]
> I didn't review and of the code changes, but I looked at the final output and the plans the compactor took to get there.

Since you looked at the tests, I'm focusing on the algorithm review and only briefly looking at the tests.

File: compactor2/src/components/split_or_compact/split_compact.rs [hunk 12]
```diff
@@ -45,25 +54,45 @@ impl SplitOrCompact for SplitCompact {
...
-            // No split is needed, which means every start-level file overlaps with at most one target-level file
-            // Need to limit number of files to compact to stay under compact size limit
-            let (files_to_compact, files_to_keep) =
-                limit_files_to_compact(self.max_compact_size, files_not_to_split, target_level);
-
-            (
+            );
+        }
+
+        // No split is needed, which means every start-level file overlaps with at most one target-level file
```
Comment (Joe-Blount) [inline compactor2/src/components/split_or_compact/split_compact.rs:68]:
```suggestion
        // No start level split is needed, which means every start-level file overlaps with at most one target-level file
```

File: compactor2/tests/layouts/knobs.rs (hunk 5 above)
Comment (Joe-Blount) [reply compactor2/tests/layouts/knobs.rs:3453]: that would be great next time, but not necessary this time.

File: compactor2/src/components/split_or_compact/files_to_compact.rs [hunk 13]
```diff
@@ -45,7 +45,7 @@ pub fn limit_files_to_compact(
     max_compact_size: usize,
     files: Vec<ParquetFile>,
     target_level: CompactionLevel,
-) -> (Vec<ParquetFile>, Vec<ParquetFile>) {
+) -> KeepAndCompactSplit {
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/files_to_compact.rs:71]: Done

File: compactor2/src/components/split_or_compact/files_to_compact.rs (hunk 9 above)
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/files_to_compact.rs:546]: > I've been bitten by logic that adds work to do once you've fallen behind. It can make it very difficult to catch back up and makes the system latch into a bad state it can't escape

There are 2 scenarios I can think of that make the compactor fall behind:
1. Some scratches or OOMs that makes the compactor not work as expected. When it is up, it has more partitions to compact. I think this is not your concern here as we focus on solving this problem else where.
2. The nature of the loading data that causes skew and as a consequence creates large files or a small file with large time range and overlaps with many large files. This is the use case of data backfill where users load data of several minutes or hours ago. I do not know how many customers do this and how often. If the ratio of loading partitions with this use case is small, then we only need to catch up with those partitions which I think very doable and not much of a concern. If there happens to be a lot of them like that, we need to rethink about our design which focuses on non-often-backfill data

File: compactor2/src/components/split_or_compact/files_to_compact.rs (hunk 10 above)
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/files_to_compact.rs]: While searching for this typo, I have found a bunch and fixed them 😄

File: compactor2/src/components/split_or_compact/large_files_to_split.rs (hunk 11 above)
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/large_files_to_split.rs]: Good point. I have added a constant of 10% for this.
I also updated the test setup of `test_compact_too_large_to_compact` to test this change

**Comment by NGA-TRAN** [comment]
@Joe-Blount I have addressed all your comments

**Review by Joe-Blount — APPROVED** [review]
LGTM, thanks!

File: compactor2/src/components/split_or_compact/files_to_compact.rs (hunk 9 above)
Comment (NGA-TRAN) [reply compactor2/src/components/split_or_compact/files_to_compact.rs:546]: Agreed

File: compactor2/tests/layouts/large_files.rs (hunk 7 above)
Comment (alamb) [reply compactor2/tests/layouts/large_files.rs:322]: For anyone following up https://github.com/influxdata/influxdb_iox/pull/7211 adds some coverage of the scenarios that the compactor is designed to work efficiently with

File: compactor2/tests/layouts/large_files.rs (hunk 8 above)
Comment (alamb) [reply compactor2/tests/layouts/large_files.rs:303]: Ok, I think the core thing I was missing was that the input pattern the compactor is optimized for.
